src/Grpc.AspNetCore.Server/Internal/HttpContextServerCallContext.cs

```diff
@@ -294,56 +398,70 @@ private TimeSpan GetTimeout()
         }
 
         // Internal for tests
-        internal void DeadlineExceeded(object state)
+        internal async void DeadlineExceeded()
         {
-            if (!_disposed)
+            // Deadline could be raised after call has been disposed
+            if (_disposed)
             {
-                Debug.Assert(_deadlineTimer != null, "Should only be called via the timer.");
-                lock (_deadlineTimer)
-                {
-                    if (!_disposed)
-                    {
-                        Log.DeadlineExceeded(_logger, GetTimeout());
-
-                        _status = new Status(StatusCode.DeadlineExceeded, "Deadline Exceeded");
+                return;
+            }
 
-                        try
-                        {
-                            // TODO(JamesNK): I believe this sends a RST_STREAM with INTERNAL_ERROR. Grpc.Core sends NO_ERROR
-                            HttpContext.Abort();
-                        }
-                        catch (Exception ex)
-                        {
-                            Log.DeadlineCancellationError(_logger, ex);
-                        }
-                    }
-                }
+            // Request abort uses the same cancellation token
+            // Don't run deadline logic if the request has already been aborted
+            if (HttpContext.RequestAborted.IsCancellationRequested)
+            {
+                return;
             }
-        }
 
-        public void Dispose()
-        {
-            var lockTaken = false;
+            Debug.Assert(_deadlineLock != null, "Lock has not been created.");
+
+            await _deadlineLock.WaitAsync();
+
             try
             {
-                // Only take a lock if there is a deadline
-                if (_deadlineTimer != null)
+                // Re-check after lock
+                if (_disposed)
+                {
+                    return;
+                }
+
+                Log.DeadlineExceeded(_logger, GetTimeout());
+
+                var status = new Status(StatusCode.DeadlineExceeded, "Deadline Exceeded");
+
+                var trailersDestination = GrpcProtocolHelpers.GetTrailersDestination(HttpContext.Response);
```

Comment:
GetStatusDestination?

Reply:
It is used for some other trailers. It is odd in this context but it makes sense elsewhere.